packages/demo-staking/src/index.spec.ts

```diff
@@ -0,0 +1,143 @@
+    const ownerPen = await Secp256k1Pen.fromMnemonic(faucet.mnemonic);
+    const ownerAddress = ownerPen.address("cosmos");
+    const ownerClient = new SigningCosmWasmClient(httpUrl, ownerAddress, (signBytes) =>
+      ownerPen.sign(signBytes),
+    );
+
+    // a user of the contract
+    const userPen = await Secp256k1Pen.fromMnemonic(faucet.mnemonic, makeCosmoshubPath(1));
+    const userAddress = userPen.address("cosmos");
+    const userClient = new SigningCosmWasmClient(
+      httpUrl,
+      userAddress,
+      (signBytes) => userPen.sign(signBytes),
+      {
+        exec: {
+          amount: coins(5000, "ucosm"),
+          gas: "300000", // 300k, needed for unbonding
+        },
+      },
+    );
+
+    const initMsg: InitMsg = {
+      name: "Bounty",
+      symbol: "BOUNTY",
+      decimals: 3,
+      validator: validator,
+      exit_tax: exitTax,
+      min_withdrawl: "7",
+    };
+    const { contractAddress } = await ownerClient.instantiate(
+      codeId,
+      initMsg,
+      `Staking derivative BOUNTY ${new Date()}`,
+    );
+
+    // Query token info (immutable)
+    {
+      const query: QueryMsg = { token_info: {} };
+      const response: TokenInfoResponse = await ownerClient.queryContractSmart(contractAddress, query);
+      expect(response).toEqual({ decimals: 3, name: "Bounty", symbol: "BOUNTY" });
+    }
+
+    // Query investment info (changes with bonding/unbonding)
+    {
+      const query: QueryMsg = { investment: {} };
+      const response: InvestmentResponse = await ownerClient.queryContractSmart(contractAddress, query);
+      expect(response).toEqual({
+        token_supply: "0",
+        staked_tokens: { denom: "ustake", amount: "0" },
+        nominal_value: "1",
+        owner: ownerAddress,
+        exit_tax: exitTax,
+        validator: validator,
+        min_withdrawl: "7",
+      });
+    }
+
+    const bondMsg: HandleMsg = { bond: {} };
+    const amount: Coin = {
+      amount: "112233",
+      denom: "ustake",
+    };
+    await userClient.execute(contractAddress, bondMsg, undefined, [amount]);
+
+    // Investment info changed
+    {
+      const quer: QueryMsg = { investment: {} };
+      const response: InvestmentResponse = await ownerClient.queryContractSmart(contractAddress, quer);
+      expect(response).toEqual({
+        token_supply: "112233",
+        staked_tokens: { denom: "ustake", amount: "112233" },
+        nominal_value: "1",
+        owner: ownerAddress,
+        exit_tax: exitTax,
+        validator: validator,
+        min_withdrawl: "7",
+      });
+    }
+
+    // Get balance
+    {
+      const query: QueryMsg = { balance: { address: userAddress } };
+      const response: BalanceResponse = await ownerClient.queryContractSmart(contractAddress, query);
+      expect(response).toEqual({
+        balance: "112233",
+      });
+    }
+
+    const unbondMsg: HandleMsg = { unbond: { amount: "112233" } };
+    // await ownerClient.execute(contractAddress, unbondMsg);
```

Comment:
@ethanfrey uncommenting this line, i.e. unbonding from an account that did not bond leads to some fun panic:

```
Error when posting tx B68E0BCC2A8A3018E43A9AA1E4A83DC4527286820E4EE5CFD8B95A561C4540EC. Code: 4; Raw log: 
  github.com/CosmWasm/wasmd/x/wasm/internal/keeper.Keeper.Execute
        github.com/CosmWasm/wasmd@/x/wasm/internal/keeper/keeper.go:198
  github.com/CosmWasm/wasmd/x/wasm.handleExecute
        github.com/CosmWasm/wasmd@/x/wasm/handler.go:104
  github.com/CosmWasm/wasmd/x/wasm.NewHandler.func1
        github.com/CosmWasm/wasmd@/x/wasm/handler.go:35
  github.com/cosmos/cosmos-sdk/baseapp.(*BaseApp).runMsgs
        github.com/cosmos/cosmos-sdk@v0.38.3/baseapp/baseapp.go:669
  github.com/cosmos/cosmos-sdk/baseapp.(*BaseApp).runTx
        github.com/cosmos/cosmos-sdk@v0.38.3/baseapp/baseapp.go:638
  github.com/cosmos/cosmos-sdk/baseapp.(*BaseApp).DeliverTx
        github.com/cosmos/cosmos-sdk@v0.38.3/baseapp/abci.go:207
  github.com/tendermint/tendermint/abci/client.(*localClient).DeliverTxAsync
        github.com/tendermint/tendermint@v0.33.3/abci/client/local_client.go:88
  github.com/tendermint/tendermint/proxy.(*appConnConsensus).DeliverTxAsync
        github.com/tendermint/tendermint@v0.33.3/proxy/app_conn.go:73
  github.com/tendermint/tendermint/state.execBlockOnProxyApp
        github.com/tendermint/tendermint@v0.33.3/state/execution.go:293
  github.com/tendermint/tendermint/state.(*BlockExecutor).ApplyBlock
        github.com/tendermint/tendermint@v0.33.3/state/execution.go:131
  github.com/tendermint/tendermint/consensus.(*State).finalizeCommit
        github.com/tendermint/tendermint@v0.33.3/consensus/state.go:1431
  github.com/tendermint/tendermint/consensus.(*State).tryFinalizeCommit
        github.com/tendermint/tendermint@v0.33.3/consensus/state.go:1350
  github.com/tendermint/tendermint/consensus.(*State).enterCommit.func1
        github.com/tendermint/tendermint@v0.33.3/consensus/state.go:1285
  github.com/tendermint/tendermint/consensus.(*State).enterCommit
        github.com/tendermint/tendermint@v0.33.3/consensus/state.go:1322
  github.com/tendermint/tendermint/consensus.(*State).addVote
        github.com/tendermint/tendermint@v0.33.3/consensus/state.go:1822
  github.com/tendermint/tendermint/consensus.(*State).tryAddVote
        github.com/tendermint/tendermint@v0.33.3/consensus/state.go:1642
  github.com/tendermint/tendermint/consensus.(*State).handleMsg
        github.com/tendermint/tendermint@v0.33.3/consensus/state.go:709
  github.com/tendermint/tendermint/consensus.(*State).receiveRoutine
        github.com/tendermint/tendermint@v0.33.3/consensus/state.go:660
  execute wasm contract failed: %!v(PANIC=Error method: unknown error variant): failed to execute message; message index: 0
```

Reply:
Huh, definitely looks like a bug, maybe in the staking contract logic, maybe in the go side.

Let me fix this before 0.8

Reply:
Fixed in https://github.com/CosmWasm/go-cosmwasm/pull/89